moto/ec2/models/instances.py

```diff
@@ -367,18 +368,30 @@ def terminate(self):
 
         self.teardown_defaults()
 
-        if self._spot_fleet_id:
-            spot_fleet = self.ec2_backend.get_spot_fleet_request(self._spot_fleet_id)
-            for spec in spot_fleet.launch_specs:
+        if self._spot_fleet_id or self._fleet_id:
+            fleet = self.ec2_backend.get_spot_fleet_request(self._spot_fleet_id)
+            if not fleet:
+                fleet = self.ec2_backend.get_fleet(
+                    self._spot_fleet_id
+                ) or self.ec2_backend.get_fleet(self._fleet_id)
+            for spec in fleet.launch_specs:
                 if (
                     spec.instance_type == self.instance_type
                     and spec.subnet_id == self.subnet_id
                 ):
                     break
-            spot_fleet.fulfilled_capacity -= spec.weighted_capacity
-            spot_fleet.spot_requests = [
-                req for req in spot_fleet.spot_requests if req.instance != self
+            fleet.fulfilled_capacity -= spec.weighted_capacity
+            fleet.spot_requests = [
+                req for req in fleet.spot_requests if req.instance != self
             ]
+            try:
+                fleet.on_demand_instances = [
+                    inst
+                    for inst in fleet.on_demand_instances
+                    if inst["instance"] != self
+                ]
+            except AttributeError:
```

Comment:
This doesn't look right to me - in what instance would this assignment throw an AttributeError?
And why can we just ignore that?

Reply:
This is because the attribute `on_demand_instances` is only present on a `fleet` resource, `spot_fleet_request` resource doesn't have on demand instances. Yea, it doesn't look right.. I can change to a _if statement_ checking if the `id` starts with `fleet` if it's better..